docs(adding-package-manager): add missing functions of ManagerAPI interface (#21535)
renovatebot/renovate: docs/development/adding-a-package-manager.md

The manager's `index.ts` file supports the following values/functions:

> - bumpPackageVersion (optional)
> - extractPackageFile
> - extractAllPackageFiles
> - getRangeStrategy (optional)
> - language (optional)
> - supportsLockFileMaintenance (optional)
> - updateArtifacts (optional)
> - updateDependency (optional)
> - updateLockedDependency (optional)

Reviewer: ```suggestion
| Value/function                | Optional |
| ----------------------------- | -------- |
| `bumpPackageVersion`          | yes      |
| `extractPackageFile`          |          |
| `extractAllPackageFiles`      |          |
| `getRangeStrategy`            | yes      |
| `language`                    | yes      |
| `supportsLockFileMaintenance` | yes      |
| `updateArtifacts`             | yes      |
| `updateDependency`            | yes      |
| `updateLockedDependency`      | yes      |
```

Changes:

- `monospace` values/functions
- Convert list to table

### `bumpPackageVersion` (optional)

Use this function to allow version bumps of updated packages.
For example, increase the version of a Maven module if a package has been updated.

> Another example would be to bump the Helm chart version, if a sub chart version has been updated.

Reviewer: ```suggestion
Another example would be to bump the Helm chart version, if a subchart version has been updated.
```

The official Helm docs use `subchart` or `subcharts`, so let's follow that convention. 😄

> Use this function if you want to allow the version bumping of updated packages.
> E.g. increase the version of a Maven module if a package has been updated.
> Another example would be to bump the Helm chart version, if a sub chart version has been updated.

Reviewer: ```suggestion
Use this function to allow version bumps of updated packages.
For example, increase the version of a Maven module if a package has been updated.
Another example would be to bump the Helm chart version, if a sub chart version has been updated.
```

The manager's `index.ts` file supports the following values/functions:

> | Value/function                | Optional |

Reviewer: Do we want a column for the `async` property? Then we list all the properties in the table.

Author: 🤔  makes sense

| ----------------------------- | -------- | ----- |
| `bumpPackageVersion`          | yes      |       |
| `extractPackageFile`          |          | yes   |

> | `extractAllPackageFiles`      |          | yes   |

Reviewer: ```suggestion
| `extractAllPackageFiles`      | yes      | yes   |
```

The heading for this value/function says:

> ### `extractAllPackageFiles(packageFiles)` (async, optional)

The table should match, I think?

Author: The first column is optional and the second async.

| `getRangeStrategy`            | yes      |
| `language`                    | yes      |
| `supportsLockFileMaintenance` | yes      |
| `updateArtifacts`             | yes      |
| `updateDependency`            | yes      |

> | `updateLockedDependency`      | yes      |

Reviewer: How do we want to order the table and the headings:

- Strict `A-Z`: messes up `optional` column
- `aync` then `normal` functions: messes op the `optional` column
- `optional:yes` first, then by `A-Z`: causes a break in the `A-Z` ordering, but keeps optional things together
- Keep things as they are now

Author: Let's keep strict A-Z. As we leave optional empty if it is false, required functions are easy identifiable.

Reviewer: I'll let you move the heading and the table items and create the `async` column. 😉

Author: I have decided to keep the order as only `extractPackageFile` and `extractAllPackageFiles` would be switched and we prefer the first one.

Reviewer: That's okay!

### `bumpPackageVersion` (optional)

Use this function to allow version bumps of updated packages.

> For example, increase the version of a Maven module if a package has been updated.

Reviewer: ```suggestion
For example, to increase the version of a Maven module if a package has been updated.
```

### `supportsLockFileMaintenance` (optional)

Set to true if this package manager needs to update lock files in addition to package files.

### `updateArtifacts` (async, optional)

This function is triggered if a dependency has been updated for a package file or `lockfileMaintenance` has been triggered.

> Mostly this function is used to directly or indirectly ( e.g. by executing external binaries ) update lock files.

Reviewer: ```suggestion
Mostly this function is used to directly (or indirectly) update lock files.
An example of indirectly updating the lock file: running external binaries.
```

How about something like this? I'm trying to split the paragraph into smaller parts.

### `updateDependency` (optional)

`updateDependency` is used if the manager can not be updated with the standard replacing mechanisms and a custom replacement has to be provided.

### `updateLockedDependency` (optional)

> ???

Author: @rarkins @viceice 
Not sure what the scope exactly is, but seems we are calling this if we are able to update exactly a specific locked dependency?

Reviewer 2: i think this is directly manipulating the lockfile. 🤔

Author: In that case we should move the Terraform lockfile support to `updateLockedDependency`, but not sure if it is triggered for `lockfileMaintenance`.

### `supportsLockFileMaintenance` (optional)

Set to true if this package manager needs to update lock files in addition to package files.

### `updateArtifacts` (async, optional)

> This function is triggered if a dependency has been updated for a package file or `lockfileMaintenance` has been triggered.

Reviewer: ```suggestion
This function triggers:

- after a dependency update (for a package file), or
- after `lockfileMaintenance`
```

This reads better to me, but it may be technically wrong. Please check before applying my suggestion. 😉

> Mostly this function is used to directly (or indirectly) update lock files.

Reviewer: ```suggestion

This function is mostly used to directly (or indirectly) update lock files.
```

### `supportsLockFileMaintenance` (optional)

Set to true if this package manager needs to update lock files in addition to package files.

### `updateArtifacts` (async, optional)

> This function is used to run binaries which in turn then will update files.

Reviewer: ```suggestion
Use `updateArtifacts` to run binaries that in turn will update files.
```

Mostly the functionality is used to indirectly update lock files.

In case of directly updating dependencies in lock files use `updateLockedDependency` instead.

> `updateArtifacts` gets triggers:

Author: `updateArtifacts` get executed on the occasion below. This would flip the causality, right? 🤔

> `updateArtifacts` gets triggers:

Reviewer: > `updateArtifacts` get executed on the occasion below. This would flip the causality, right? 🤔

Do you mean that _something else_ triggers the `updateArtifacts` function, and thus `triggers` is wrong?

How about `gets triggered` to make clear _something else_ triggers the function:

```suggestion
`updateArtifacts` gets triggered:
```

An example of indirectly updating the lock file: running external binaries.

> In case of directly updating dependencies in lock files use `updateLockedDependency` instead.

Reviewer: This is still a bit confusing to me. It looks like we say:

1. Use `updateArtifacts` to directly/indirectly update lock files
2. Use `updateLockedDependency` to update dependencies in lockfiles

I assume that when we update a lockfile we always update the dependencies? I mean if you're not updating dependencies, why touch the lockfile? 😕

Author: Contributors should choose `updateLockedDependency` if they are directly manipulating lock files. 

`updateArtifacts` if they are doing it indirectly e.g. by using an external binary or other files.     
But legacy implementations still directly edit lock files here, but we want to move away from that.

Reviewer: Can you put your expanded explanation in the docs? I'll review the text and make improvements if needed. 😉

### `updateDependency` (optional)

> `updateDependency` is used if the manager can not be updated with the standard replacing mechanisms and a custom replacement has to be provided.

Reviewer: ```suggestion
Use `updateDependency` if _both_ conditions apply:

- the manager can't be updated to use the standard replacing mechanism
- a custom replacement has to be provided
```

My rewrite still feels clunky. At least we have a list now. 😄

### `updateLockedDependency` (optional)

> This function should be used if dependencies in lock files are directly updated.

Reviewer: ```suggestion
Use `updateLockedDependency` to directly update dependencies in lock files.
```
